src/lib/aggregator.ts

```diff
@@ -1,58 +1,144 @@
 import {STRINGS} from '../config';
 
-import {PluginParams} from '../types/interface';
-import {Parameters} from '../types/parameters';
-import {AggregationResult} from '../types/aggregation';
-
+import {ManifestCommon} from '../types/manifest';
+import {Node} from '../types/compute';
 const {InvalidAggregationParams} = ERRORS;
-const {INVALID_AGGREGATION_METHOD, METRIC_MISSING} = STRINGS;
+const {INVALID_AGGREGATION_METHOD} = STRINGS;
 
 /**
  * Validates metrics array before applying aggregator.
  * If aggregation method is `none`, then throws error.
  */
-const checkIfMetricsAreValid = (metrics: string[], parameters: Parameters) => {
-  metrics.forEach(metric => {
-    const method = getAggregationMethod(metric, parameters);
-
+const checkIfMetricsAreValid = (context: ManifestCommon) => {
+  context.aggregation?.metrics.forEach(metric => {
+    const method = getAggregationMethod(metric, PARAMETERS);
     if (method === 'none') {
       throw new InvalidAggregationParams(INVALID_AGGREGATION_METHOD(method));
     }
   });
 };
 
-/**
- * Aggregates child node level metrics. Validates if metric aggregation type is `none`, then rejects with error.
- * Otherwise iterates over inputs by aggregating per given `metrics`.
- */
-export const aggregate = (
-  inputs: PluginParams[],
-  metrics: string[],
-  parameters: Parameters
-) => {
-  checkIfMetricsAreValid(metrics, parameters);
-
-  return inputs.reduce((acc, input, index) => {
-    for (const metric of metrics) {
-      if (!(metric in input)) {
-        throw new InvalidAggregationParams(METRIC_MISSING(metric, index));
+/**Returns Boolean indicating whether a node has children */
+function hasChildren(node: Node) {
+  if (node['children']) {
+    return true;
+  }
+  return false;
+}
+
+/** Returns children of given node */
+function getChildren(node: Node): Node[] {
+  const children: Node[] = [];
+  Object.keys(node.children).forEach(child => {
+    children.push(node.children[child]);
+  });
+  return children;
+}
+
+/** element-wise addition of two arrays */
+function aggregateArrays(a: Array<any>, b: Array<any>, method: string) {
+  if (method === 'sum') {
+    return a.map((element, idx) => {
+      return element + b[idx];
+    });
+  } else if (method === 'avg') {
+    return a.map((element, idx) => {
+      return (element + b[idx]) / 2;
+    });
+  }
+  return a;
+}
+
+/** aggregates multiple time series into one single time series*/
+export const aggregate = (tree: Node, context: ManifestCommon) => {
+  checkIfMetricsAreValid(context);
+  // check if root has children
+  if (hasChildren(tree)) {
+    // iterate through child nodes
+    getChildren(tree).forEach(node => {
+      // if the child has children, start recursive call to aggregate()
+      if (hasChildren(node)) {
+        aggregate(node, context);
       }
+      context.aggregation?.metrics.forEach(metric => {
+        let treeTimeSeries = tree.inputs?.map(input => input[`${metric}`]);
```

Comment:
So tree here is referring to the parent node of the current node.

Reply:
yes, you can swap "tree" for "root" if it's easier to reason about

Reply:
agree that the ;logic will be much easier if the parent cannot also have its own input data